modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java

```diff
@@ -0,0 +1,365 @@
+            },
+            failure -> {
+                LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to download database [{}]", databaseName), failure);
+                try {
+                    Files.delete(databaseTmpFile);
+                    Files.delete(databaseTmpGzFile);
+                } catch (IOException ioe) {
+                    ioe.addSuppressed(failure);
+                    LOGGER.error("Unable to delete tmp database file after failure", ioe);
+                }
+            });
+    }
+
+    void updateDatabase(String databaseFileName, String recordedMd5, Path file) {
+        try {
+            LOGGER.info("database file changed [{}], reload database...", file);
+            DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, file);
+            DatabaseReference existing = databases.put(databaseFileName, new DatabaseReference(recordedMd5, file, loader));
+            if (existing != null) {
+                existing.close();
+            }
+        } catch (Exception e) {
+            LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to update database [{}]", databaseFileName), e);
+        }
+    }
+
+    void removeStaleEntries(Collection<String> staleEntries) {
+        try {
+            List<DatabaseReference> references = new ArrayList<>();
+            for (String staleEntry : staleEntries) {
+                LOGGER.info("database [{}] no longer exists, cleaning up...", staleEntry);
+                DatabaseReference existing = databases.remove(staleEntry);
+                assert existing != null;
+                references.add(existing);
+            }
+
+            for (DatabaseReference reference : references) {
+                reference.close();
+                Files.delete(reference.databaseFile);
+            }
+        } catch (Exception e) {
+            LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to clean database [{}]", staleEntries), e);
+        }
+    }
+
+    void retrieveDatabase(String databaseName,
+                          String expectedMd5,
+                          GeoIpTaskState.Metadata metadata,
+                          CheckedConsumer<byte[], IOException> chunkConsumer,
+                          CheckedRunnable<Exception> completedHandler,
+                          Consumer<Exception> failureHandler) {
+        // Need to run the search from a different thread, since this is executed from cluster state applier thread:
+        genericExecutor.accept(() -> {
+            MessageDigest md = MessageDigests.md5();
+            Long lastSortValue = null;
+
+            try {
+                do {
+                    SearchRequest searchRequest =
+                        createSearchRequest(databaseName, metadata.getFirstChunk(), metadata.getLastChunk(), lastSortValue);
+                    if (lastSortValue != null) {
+                        lastSortValue = null;
+                    }
```

Comment:
```suggestion
                    lastSortValue = null;
```
(we always end up with `null` here anyway)

Reply:
i think that is a leftover.

Reply:
(the if check)